app/controllers/application_controller.rb

```diff
@@ -477,6 +479,28 @@ def secure_session
     SecureSession.new(session["secure_session_id"] ||= SecureRandom.hex)
   end
 
+  def set_csp_header
+    nonce = SecureRandom.base64
+    request.env["nonce"] = nonce
+    response.headers["X-Discourse-CSP-Nonce"] = nonce
```

Comment:
Expose the nonce in a special header to allow site admins to set their CSP entirely in nginx (among other things I'm sure this makes possible).